processor/src/main/java/com/linecorp/decaton/processor/runtime/internal/CompletionImpl.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.decaton.processor.runtime.internal;
+
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.function.Function;
+
+import com.linecorp.decaton.processor.runtime.Completion;
+
+import lombok.Setter;
+import lombok.experimental.Accessors;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+@Accessors(fluent = true)
+public class CompletionImpl implements Completion {
+    private final CompletableFuture<Void> future;
+    @Setter
+    private volatile Function<Completion, TimeoutChoice> expireCallback;
+    private volatile Completion dependency;
+
+    public static CompletionImpl completedCompletion() {
+        CompletionImpl comp = new CompletionImpl();
+        comp.complete();
+        return comp;
+    }
+
+    public CompletionImpl() {
+        future = new CompletableFuture<>();
+    }
+
+    @Override
+    public boolean isComplete() {
+        return future.isDone();
+    }
+
+    @Override
+    public CompletionStage<Void> asFuture() {
+        return future;
+    }
+
+    @Override
+    public boolean tryExpire() {
+        if (isComplete()) {
+            return true;
+        }
+        if (dependency != null) {
+            if (!dependency.tryExpire()) {
+                return false;
+            }
+            dependency = null;
```

Comment:
(Just a confirmation) Say we have processors like below:

```
- CompactionProcessor: deferCompletion(() -> GIVE_UP).completeWith(ctx.push())
  - UserProcessor1: deferCompletion(() -> EXTEND).completeWith(ctx.push())
    - UserPostProcessor2: deferCompletion(() -> GIVE_UP).completeWith(doSomethingOnAsyncFramework())
```

When completion timed out once, since `UserProcessor2`'s choice is GIVE_UP, `UserProcessor1`'s dependency will be substituted with null.

Since `UserProcessor1`'s choice is EXTEND, the timeout will be extended.

When the timeout happens 2nd time, since `UserProcessor1`'s dependency is already null, `UserProcessor2`'s callback will not be called.

But it is expected and okay, because `UserProcessor2` chose GIVE_UP on first timeout, it means that we don't have to take `UserProcessor2`'s choice into account of expiration handling.

Is my understanding correct?

Reply:
That's correct.